btf.c

```diff
@@ -192,8 +194,96 @@ btf_root_offset(struct btf *btf, const char *dotname)
 	return (btf_root_offset2(btf, dotname));
 }
 
+static struct quark_btf *
+quark_btf_new(const char *new_name)
+{
+	struct quark_btf	*qbtf;
+
+	if ((qbtf = malloc(sizeof(*qbtf) + sizeof(base_targets))) == NULL)
+		return (NULL);
+	qbtf->kname = strdup(new_name);
+	if (qbtf->kname == NULL) {
+		free(qbtf);
+		return (NULL);
+	}
+	memcpy(qbtf->targets, base_targets, sizeof(base_targets));
+
+	return (qbtf);
+}
+
+static struct quark_btf *
+quark_btf_dup(struct quark_btf *src)
+{
+	struct quark_btf	*qbtf;
+
+	if (src == NULL)
+		return (NULL);
+
+	qbtf = quark_btf_new(src->kname);
+	if (qbtf == NULL)
+		return (NULL);
+	memcpy(qbtf->targets, src->targets, sizeof(base_targets));
+
+	return (qbtf);
+}
+
 struct quark_btf *
-quark_btf_open(const char *path, const char *kname)
+quark_btf_open_hub(const char *version)
+{
+	extern struct quark_btf	 *all_btfs[];
+	struct quark_btf	**pp, *best, *cand;
+	int			  best_score;
+
+	/* paranoia */
+	if (version == NULL || strlen(version) == 0)
+		return (NULL);
+
+	best = NULL;
+	best_score = 0;
+	for (pp = all_btfs; (cand = *pp) != NULL; pp++) {
+		const char	*pv, *pc;
+		int		 score;
+
+		score = 0;
+
+		/* paranoia */
+		if (cand->kname == NULL || strlen(cand->kname) == 0)
+			return (NULL);
+
+		/*
+		 * Match head, the beginning of version
+		 */
+		for (pc = cand->kname, pv = version, score = 0;
+		     *pc != 0 && *pv != 0 && *pc == *pv;
+		     score++, pc++, pv++)
+			;     /* NADA */
+
+		/*
+		 * If we didn't score yet, don't bother matching tail
+		 */
+		if (score == 0)
+			continue;
+
+		/*
+		 * Match tail, the end of version
+		 */
+		for (pc = cand->kname + strlen(cand->kname) - 1,
```

Comment:
That's just pointer arithmetics, it's the equivalent of
`pc = cand->kname[strlen(cand->kname) - 1]`
Because the [] operator is just sugar for `*(a + b)`
`a[b] = 1` is `*(a + b) = 1`

Reply:
actually for consistency, the `score` in the third for should be in the end, I'll change it, so it's consistent

Reply:
> Equally spooky. Assigning the first character of a string plus the length of that same string 🤯

Maybe I misunderstood the spooky part, you can't really get the address of the last character of a string without knowing long it is. Do you mean the substraction is scary? cause cand->kname[-1] would be very evil, that's why there is a check in the beginning.